src/overlays/actors/ovl_En_Dog/z_en_dog.c

```diff
@@ -22,41 +28,579 @@ const ActorInit En_Dog_InitVars =
+    func_80034EC0(skelAnime, animations, 0);
+    
+    if ((this->actor.params & 0x8000) == 0)
+    {
+        this->actor.params = (this->actor.params & 0xF0FF) | ((((this->actor.params & 0x0F00) >> 8) + 1) << 8);
+    }
+
+    followingDog = ((gSaveContext.dogParams & 0x0F00) >> 8);
+    if (followingDog == ((this->actor.params & 0x0F00) >> 8) && ((this->actor.params & 0x8000) == 0))
+    {
+        Actor_Kill(&this->actor);
+        return;
+    }
+
+    ActorCollider_AllocCylinder(globalCtx, collider);
+    ActorCollider_InitCylinder(globalCtx, collider, &this->actor, &cylinderInit);
+    func_80061EFC(&this->actor.sub_98, 0, &sub98Data);
+    Actor_SetScale(&this->actor, 0.0075f);
+    this->waypoint = 0;
+    this->actor.gravity = -1.0f;
+    this->path = func_8008E4E0(globalCtx, (s16)((this->actor.params & 0x00F0) >> 4), 0xF);
+
+    switch (globalCtx->sceneNum)
+    {
+        case SCENE_MARKET_NIGHT:
+            if ((gSaveContext.richardIsLost == 0) && (((this->actor.params & 0x0F00) >> 8) == 1))
+            {
+                Actor_Kill(&this->actor);
+            }
+            break;
+        // Richard's Home
+        case SCENE_IMPA:
+            if ((u32)(this->actor.params & 0x8000) == 0)
+            {
+                if (gSaveContext.richardIsLost == 0)
+                {
+                    this->nextBehavior = DOG_SIT;
+                    this->actionFunc = EnDog_Wait;
+                    this->actor.speedXZ = 0.0f;
+                    return;
+                }
+                else
+                {
+                    Actor_Kill(&this->actor);
+                    return;
+                }
+            }
+            break;
+
+        default: break;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FB3AC.s")
+    if ((u32)(this->actor.params & 0x8000) != 0)
+    {
+        this->nextBehavior = DOG_WALK;
+        this->actionFunc = EnDog_FollowLink;
+    }
+    else
+    {
+        this->nextBehavior = DOG_SIT;
+        this->actionFunc = EnDog_ChooseMovement;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/EnDog_Init.s")
+static void EnDog_Destroy(EnDog* this, GlobalContext* globalCtx)
+{
+    ColliderCylinderMain* collider = &this->collider; 
+    ActorCollider_FreeCylinder(globalCtx, collider);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/EnDog_Destroy.s")
+static void EnDog_FollowPath(EnDog* this, GlobalContext* globalCtx)
+{
+    s16ArrEntry behaviors[] = { 3, 5, 2 };
+    s16ArrEntry unused[] = { 40, 80, 20 };
+    f32 speed;
+    s32 frame;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FB6C4.s")
+    if (EnDog_CanFollow(this, globalCtx) == 1)
+    {
+        this->actionFunc = EnDog_FollowLink;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FB858.s")
+    if(DECR(this->behaviorTimer))
```

Comment:
Here it's not a boolean check so I would actually keep the != 0
```suggestion
    if (DECR(this->behaviorTimer) != 0)
```